runtime/src/gpu/rocm/gpu-rocm.c

```diff
@@ -27,67 +27,497 @@
+  // object to be on the GPU).
+  //
+  // It's a bit of a hack but to handle this we start off setting
+  // `chpl_gpu_device_alloc` to false indicating that we shouldn't actually
+  // do any allocations on the device. Once the standard modules have finished
+  // loading this callback function
+  // (`chpl_gpu_impl_on_std_modules_finished_initializing`) gets called and we
+  // flip the flag.
+  chpl_gpu_device_alloc = true;
+}
 
 void chpl_gpu_get_device_count(int* into) {
   hipGetDeviceCount(into);
 }
 
-void chpl_gpu_impl_launch_kernel(int ln, int32_t fn,
-                                 const char* name,
-                                 int grd_dim_x, int grd_dim_y, int grd_dim_z,
-                                 int blk_dim_x, int blk_dim_y, int blk_dim_z,
-                                 int nargs, va_list args) {}
+bool chpl_gpu_impl_is_device_ptr(const void* ptr) {
+  hipPointerAttribute_t res;
+  hipError_t ret_val = hipPointerGetAttributes(&res, (hipDeviceptr_t)ptr);
 
-void chpl_gpu_impl_launch_kernel_flat(int ln, int32_t fn,
-                                      const char* name,
-                                      int num_threads, int blk_dim, int nargs,
-                                      va_list args) {}
+  if (ret_val != hipSuccess) {
+    if (ret_val == hipErrorInvalidValue ||
+        ret_val == hipErrorNotInitialized ||
+        ret_val == hipErrorDeinitialized) {
+      return false;
+    }
+    else {
+      CUDA_CALL(ret_val);
+    }
+  }
 
-void* chpl_gpu_mem_array_alloc(size_t size, chpl_mem_descInt_t description,
-                               int32_t lineno, int32_t filename) {
-  return chpl_malloc(size);
+  #ifdef CHPL_GPU_MEM_STRATEGY_ARRAY_ON_DEVICE
+    return res.memoryType == hipMemoryTypeDevice;
+  #else
+    return true;
+  #endif
 }
 
-void* chpl_gpu_impl_mem_alloc(size_t size) {
-  return chpl_malloc(size);
+bool chpl_gpu_impl_is_host_ptr(const void* ptr) {
+  hipPointerAttribute_t res;
+  hipError_t ret_val = hipPointerGetAttributes(&res, (hipDeviceptr_t)ptr);
+
+  if (ret_val != hipSuccess) {
+    if (ret_val == hipErrorInvalidValue ||
+        ret_val == hipErrorNotInitialized ||
+        ret_val == hipErrorDeinitialized) {
+      return true;
+    }
+    else {
+      CUDA_CALL(ret_val);
+    }
+  }
+  else {
+    return res.memoryType == hipMemoryTypeHost;
+  }
+
+  return true;
+}
+
+static void chpl_gpu_launch_kernel_help(int ln,
+                                        int32_t fn,
+                                        const char* name,
+                                        int grd_dim_x,
+                                        int grd_dim_y,
+                                        int grd_dim_z,
+                                        int blk_dim_x,
+                                        int blk_dim_y,
+                                        int blk_dim_z,
+                                        int nargs,
+                                        va_list args) {
```

Comment:
Definitely not for this PR, but I'd like this function to be a much simpler wrapper around the interface we are using, where this logic is in `runtime/src/chpl-gpu.c`.

I didn't do that because we didn't know whether the logic will be common between NVIDIA and AMD. The use of HIP is still a little bit open, but if that's the direction we are adopting, we should definitely avoid this redundancy.